### PR TITLE
Make JLanguageHelper::getInstalledLanguages add log message when language xml files doesn't exist (or is not readable)

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for New Issue.

### Summary of Changes

When the language xml files does not exist the new `JLanguageHelper::getInstalledLanguages` is returning an exception. This could cause some B/C issues.
So this PR makes it fail silently instead.

### Testing Instructions

1. Have a site with 2 or more languages installed
2. Delete (or rename) one site language xml file (in `/languages/xx-XX/xx-XX.xml`)
3. Go to Extensions -> Languages -> Installed. You will see an exception being trowed
4. Apply patch
5. Repeat step 3, no exception and you will see a log message in debug console

### Documentation Changes Required

None.

### CODE REVIEW

**Comment by brianteeman** [comment]
Is it really a good idea to ever suppress errors

**Comment by andrepereiradasilva** [comment]
The thing is: i have a PR (https://github.com/joomla/joomla-cms/pull/12671) for this method to be used across the core (including frontend) and as it is if somehow the language xml is not readble it will trow an exception. Before it wouldn't, so to preserve the same behaviour i made this PR.

**Comment by ggppdk** [comment]
Yes catch the exception and avoid the meaningless fatal error, but issue no message at all ?

why not replace the fatal meaningless error with a good warning message if JApplication is backend, instead of ignoring it silently ?

**Comment by mbabker** [comment]
It's consistent with other checks in that part of the code, seems fine to me.  At most I'd suggest a log message but I don't see the need to display a message in the UI regardless of client.

**Comment by andrepereiradasilva** [comment]
ok. added log messages

**Comment by andrepereiradasilva** [comment]
humm ... this way i also get warning messages (?)

**Comment by mbabker** [comment]
Logging to the `jerror` category causes this.  Log to any other category and it won't automagically get added to the message queue.

**Comment by andrepereiradasilva** [comment]
> Logging to the jerror category causes this. Log to any other category and it won't automagically get added to the message queue.

hum ... are there other categories besides `database, databasequery, database-error, deprecated and jerror`?

**Comment by mbabker** [comment]
You can use whatever you want for the category.  Those are just the
standard values we use in core.

**Comment by andrepereiradasilva** [comment]
oh ok, learning every day :wink:

**Comment by infograf768** [comment]
I agree with @ggppdk 
I would prefer to get a good old Error message in back-end. That would be real useful for admins.

**Comment by andrepereiradasilva** [comment]
the thing is the old message was an individual check in com_languages installed model that was making an hartdcoded query to fetch the installed languages.

That individual check was removed since the languages are now fetched from JLanguageHeler and not an hardcoded individual  

I ask you to see the global perspective here, not a particular place where the warning message appear (by the way it was a PR of mine that putted there before :wink: ).

The installed languages currently are checked (db extension, file, metadata) is some places and other not. The end goal of moving all this to JLanguageHelper is to avoid code repetition and with that perfomance/memory improvings and especcialy correct/avoid errors and unexpected behaviours.

Besides this PR and the one already merged (https://github.com/joomla/joomla-cms/pull/12641) i have made already another PR (https://github.com/joomla/joomla-cms/pull/12671) to use always this new method, so we can't send the warning messages everywhere the instralled languages are fetch, instead we now log them to debug console.

In summary, you can still see the warning messages if you enable debug and system debug plugin and go to log messages tab. 

I also plan to do something similiar with the content languages in the future.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 2a27a735825d8661d2b580f6bf92ca22a2fa40b6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12783">issues.joomla.org/tracker/joomla-cms/12783</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@alikon can you test this one too?

**Comment by andrepereiradasilva** [comment]
could we get another test here.
Without this and since https://github.com/joomla/joomla-cms/pull/12671 was already merged now we have exceptions in staging if you have a language installed without the language files.

**Review by mbabker — APPROVED** [review]

**Comment by Bakual** [comment]
Tested fine. However I wasn't able to get the log message appear. Personally I would prefer a warning message as well, since there is something wrong in the system for sure if that case happens.

Also I'm not sure why you try/catch the `parseXMLInstallFile` call. That one returns a array|boolean and doesn't throw anything from what I see.

**Comment by mbabker** [comment]
> Also I'm not sure why you try/catch the parseXMLInstallFile call. That one returns a array|boolean and doesn't throw anything from what I see.

`simplexml_load_file()` instantiates a new `SimpleXMLElement` object and its constructor can throw an `Exception`.

**Comment by andrepereiradasilva** [comment]
@Bakual can you mark as tested with success?

**Comment by mbabker** [comment]
> Tested fine. However I wasn't able to get the log message appear. Personally I would prefer a warning message as well, since there is something wrong in the system for sure if that case happens.

This is a high level library class.  So this shouldn't be directly attached to the application classes or the message queue.  Next, this error isn't appropriate for the frontend if this API ever gets used there as it would include an information disclosure because file paths are being used, so if you wanted to enqueue messages, now you'd have to have application specific behaviors in a high level library class (even worse, I'm still annoyed by that in the Redis cache adapter).  So that means letting the error bubble and the callers need to catch errors, and lets face it, people don't do error handling in Joomla (even the error handling in core is barely a step above bad).

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on 2a27a735825d8661d2b580f6bf92ca22a2fa40b6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12783">issues.joomla.org/tracker/joomla-cms/12783</a>.</sub>

**Comment by dgt41** [comment]
RTC

**Comment by Bakual** [comment]
> This is a high level library class. So this shouldn't be directly attached to the application classes or the message queue.

Yes sure. Throwing the exception in JLanguage is fine. Catching that and enqueueing a message in the JLanguageHelper however is no issue as that class has several couplings to the application.
As for frontend, the helper class can (and does) distinguish it. So the warning could only be shown in backend. And obviously I wouldn't show the filepath. It isn't needed anyway for a helpful message.
See my PR (which I did without knowing of this one): #12921

> can you mark as tested with success?
@andrepereiradasilva No, because I wasn't able to see the log message, disagree with only logging and had another question. So it's not a successfull (but also not really failed) test for me.

**Comment by andrepereiradasilva** [comment]
if you don't see a log message is because you have the cache enabled (the cache issue will be solved with https://github.com/joomla/joomla-cms/pull/12906) or you didn't enable the global config debug and debug plugin to show those log messages.

**Comment by mbabker** [comment]
Just because we already have bad architecture doesn't mean we should add more.

> Catching that and enqueueing a message in the JLanguageHelper however is no issue as that class has several couplings to the application.

No it doesn't.  It actually has one and that's for the install app.  The rest of the helper class is (correctly) application ignorant.  This isn't to be confused with the client ID parameter in the `getInstalledLanguages()` which isn't a coupling to the application; it's a parameter to filter data, not something that requires use of the application object.

> So the warning could only be shown in backend.

Anything in the libraries folder which is making decisions to enqueue a message based on the active application has a major code smell.  Message enqueuing should not be so liberally used everywhere as that creates dependencies to the application object, and inherently JFactory, and if anyone ever actually gave the test suite the proper attention you'd realize what kind of nightmare that creates because of how much of the global state you have to mock just to be able to run a test condition.

> It [file path] isn't needed anyway for a helpful message.

For debugging missing file errors, it actually is pretty helpful.  Unless you're assuming the message can be written in a way that someone debugging it can immediately decipher what the cause of the issue was.  Your PR honestly just adds another vague error message to be displayed.

> I wasn't able to see the log message

Ensure you have a logger set up catching the "language" category.  Since Joomla by default doesn't turn on any loggers except to proxy the "jerror" category to the message queue it won't be logged without turning on additional tools.

**Comment by Bakual** [comment]
Found why the log didn't work. When testing this PR with the patchtester, the log wasn't created for some reason. There were also other changes than the ones from this PR. That's due to the way the Patchtester works (replacing files instead of merging changes).

When merging directly from you branch, the log appeared when the debug plugin had "Log Almost Everything" enabled as expected. Btw the global debug setting doesn't play a role here.

Still don't like that this is hidden in a log which nobody will see. But if I'm the only one here, feel free to merge it.

**Comment by infograf768** [comment]
You are not the only one, but, whatever... :smiley:
